<a id="item31"></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np


<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

### Normalization of data

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors
print(n_cols)

8


### Split the data into training and testing sets

let's import `train_test_split` from the `sklearn`

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

<a id="item1"></a>

<a id='item32'></a>

## Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

#### Let's go ahead and import the Keras library

In [12]:
import keras

Using TensorFlow backend.


As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [13]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that one hidden layer of 10 hidden units.

<a id="item4"></a>

<a id='item34'></a>

## Train and Test the Network

Let's call the function now to create our model.

In [15]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

In [16]:
# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 0s - loss: 1603.7927 - val_loss: 1150.6177
Epoch 2/100
 - 0s - loss: 1580.7934 - val_loss: 1137.6873
Epoch 3/100
 - 0s - loss: 1557.0257 - val_loss: 1124.1964
Epoch 4/100
 - 0s - loss: 1531.8328 - val_loss: 1110.1509
Epoch 5/100
 - 0s - loss: 1504.9854 - val_loss: 1095.2274
Epoch 6/100
 - 0s - loss: 1477.3671 - val_loss: 1079.4499
Epoch 7/100
 - 0s - loss: 1447.4419 - val_loss: 1063.3305
Epoch 8/100
 - 0s - loss: 1417.0983 - val_loss: 1046.0415
Epoch 9/100
 - 0s - loss: 1384.8569 - val_loss: 1028.4264
Epoch 10/100
 - 0s - loss: 1351.6269 - val_loss: 1009.6252
Epoch 11/100
 - 0s - loss: 1317.1078 - val_loss: 990.6822
Epoch 12/100
 - 0s - loss: 1282.1568 - val_loss: 970.4101
Epoch 13/100
 - 0s - loss: 1245.1436 - val_loss: 950.5448
Epoch 14/100
 - 0s - loss: 1208.2306 - val_loss: 929.5075
Epoch 15/100
 - 0s - loss: 1170.2500 - val_loss: 908.1382
Epoch 16/100
 - 0s - loss: 1131.2546 - val_loss: 886.6438
Epoch 17/100
 - 0s - loss

In [17]:
model.evaluate(X_test,y_test, verbose=1)

309/309 [==============================] - 0s 27us/step


154.23022628833561

   Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

   Let's import the mean_squared_error function from Scikit-learn.

In [18]:
from sklearn.metrics import mean_squared_error

In [19]:
y_pred = model.predict(X_test)

In [20]:
y_pred

array([[53.774673 ],
       [44.948288 ],
       [63.50659  ],
       [44.35504  ],
       [21.074451 ],
       [25.763252 ],
       [14.619875 ],
       [50.983826 ],
       [33.56835  ],
       [34.97419  ],
       [22.500927 ],
       [20.879028 ],
       [73.6152   ],
       [46.578354 ],
       [32.95408  ],
       [42.91245  ],
       [17.562462 ],
       [25.154303 ],
       [16.512577 ],
       [23.903444 ],
       [32.875393 ],
       [32.48267  ],
       [31.934048 ],
       [27.057    ],
       [33.187775 ],
       [37.486187 ],
       [ 9.540255 ],
       [37.174416 ],
       [50.744686 ],
       [18.781693 ],
       [36.5395   ],
       [28.302425 ],
       [44.34806  ],
       [54.398018 ],
       [22.000849 ],
       [35.906906 ],
       [32.05563  ],
       [26.55257  ],
       [19.185535 ],
       [33.26503  ],
       [14.740544 ],
       [16.23128  ],
       [24.362679 ],
       [46.270554 ],
       [15.727529 ],
       [50.874886 ],
       [53.168873 ],
       [60.91

In [21]:
mse = mean_squared_error(y_test, y_pred)
print('mse: ', mse)

mse:  154.23022948236525


Let's create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [22]:
n = 50
epochs = 100
mean_squared_errors = []
for i in range(0, n):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    mse = model.evaluate(X_test, y_test, verbose=0)
    print("mse " +str(i+1)+": ", mse)
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)
    
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)   

print("Mean : ", mean)
print("Standard Deviation : ", standard_deviation)

mse 1:  96.0993934557276
mse 2:  79.56151080825954
mse 3:  74.07896404821896
mse 4:  71.91631670214211
mse 5:  70.28211563999213
mse 6:  70.17010316570986
mse 7:  69.81546240254127
mse 8:  69.58942292346151
mse 9:  69.40679392151077
mse 10:  69.90222286483616
mse 11:  69.22785313924153
mse 12:  69.29306682105204
mse 13:  69.50471144741022
mse 14:  69.4231158753429
mse 15:  69.42239610739897
mse 16:  69.34406570632095
mse 17:  69.05443929777176
mse 18:  69.22501417734091
mse 19:  69.48181279500325
mse 20:  69.16611189363844
mse 21:  69.12804343708125
mse 22:  68.99314477914359
mse 23:  69.13348335587091
mse 24:  69.14918734109132
mse 25:  69.09194231496274
mse 26:  68.98234958093143
mse 27:  69.27858135383877
mse 28:  69.07535241876991
mse 29:  69.27334270045209
mse 30:  69.02938621792593
mse 31:  69.34685383028197
mse 32:  69.16655685523567
mse 33:  69.43408119176969
mse 34:  69.05792305461797
mse 35:  69.11284363385543
mse 36:  68.914045796811
mse 37:  69.13508775087622
mse 38:  68.89

In [23]:
print('Below are the mean and standard deviation of 50 mean squared errors of normalized data for 100 epochs')
print("Mean : ", mean)
print("Standard Deviation : ", standard_deviation)

Below are the mean and standard deviation of 50 mean squared errors of normalized data for 100 epochs
Mean :  70.15126341649227
Standard Deviation :  4.052429191995384


In [24]:
print('mean squared errors: ', mean_squared_errors)

mean squared errors:  [96.09939568 79.56151237 74.07896564 71.91631827 70.28211789 70.17010454
 69.81546261 69.58942346 69.40679577 69.90222365 69.22785449 69.29306562
 69.50471092 69.4231171  69.42239676 69.34406566 69.05443898 69.22501508
 69.48181303 69.16611277 69.12804231 68.99314616 69.13348549 69.14918872
 69.09194264 68.98234879 69.27858193 69.07535222 69.27334222 69.02938645
 69.34685332 69.16655692 69.43408289 69.05792364 69.11284311 68.91404738
 69.13508795 68.89255225 69.10776252 69.16763935 69.32058191 68.929416
 68.97798714 69.24023851 69.17798452 69.31095476 69.08296686 68.99525295
 69.1177615  68.97495013]


The `mean` with predictors_norm data of 50 mean squared errors in part B for 100 epochs is 70 while it was 51 in part B  for 50 epochs.